server/src/main/java/org/opensearch/identity/tokens/TokenManager.java

```diff
@@ -0,0 +1,47 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.identity.tokens;
+
+/**
+ * This interface defines the expected methods of a token manager
+ */
+public interface TokenManager {
```

Comment:
Looks like currently the TokenManager is designed per auth token type (`BasicAuthToken`, `NoopToken`), may be we could encode that in type itself? 

```
public interface TokenManager<T extends AuthToken>  {
 ....
}
```

Than we could get rid of typecasts, what do you think?

Reply:
So where this gets tricky is with implementations of the Identity Plugin which support multiple token types. The idea with this is to be able to support specific operations based on the token type. Like Basic or Bearer etc. But with the Security Plugin and likely other Identity Plugins, you will need to handle multiple types. The way this is setup is by using a "sorting" manager such as shown here: https://github.com/opensearch-project/security/pull/2787. Basically, there is a Plugin level sorter that identifies the type of a given token and directs it to the correct token manager. I am not sure we can do this if we encode the types since they would not have a specific type in these cases. 

Perhaps I am wrong though? Do you think that would still or work or there is a better way of making a 1:1 mapping of token manager to identity interface but also managing multiple token types?

Reply:
Fair enough, I think in this case keeping it generic is fine, thank you.